### PR TITLE
Pyethereum keystore support

### PR DESCRIPTION
**Pull request by tworec**
Continuation of #9444
(sorry for duplicate, but I can not reopen #9444)

### CODE REVIEW

File: ethstore/src/account/safe_account.rs [hunk 1]
```diff
@@ -76,17 +76,28 @@ impl SafeAccount {
...
-	/// can be left `None`.
-	pub fn from_file(json: json::KeyFile, filename: Option<String>) -> Self {
-		SafeAccount {
+	/// can be left `None`. `password` is used for reading keystore files w/o address.
+	pub fn from_file(json: json::KeyFile, filename: Option<String>, password: &Option<Password>) -> Result<Self, Error> {
+		let crypto = Crypto::from(json.crypto);
+		let address = match json.address {
+			Some(address) => address.into(),
+			None => match password {
+                Some(password) =>
```
Comment (debris) [inline ethstore/src/account/safe_account.rs]: invalid indentation

File: ethstore/src/accounts_dir/disk.rs [hunk 2]
```diff
@@ -104,8 +105,16 @@ pub type RootDiskDirectory = DiskDirectory<DiskKeyFileManager>;
 
 /// Disk directory key file manager
 pub trait KeyFileManager: Send + Sync {
+	/// Read `SafeAccount` from given key file stream with password
+	#[allow(unused_variables)]
+	fn read_with_password<T>(&self, filename: Option<String>, reader: T, password: &Option<Password>)
+		-> Result<SafeAccount, Error> where T: io::Read {
+		unimplemented!()
```
Comment (debris) [inline ethstore/src/accounts_dir/disk.rs]: this is a badly designed interface. what you should do is:

```rust
pub struct KeyFileManagerWithPassword {
  password: Password,
}

impl KeyFileManager for KeyFileManagerWithPassword {
  fn read<T>(&self, filename: Option<String>, reader: T) -> Result<SafeAccount, Error> where T: io::Read;
}
```

File: ethstore/src/accounts_dir/mod.rs [hunk 3]
```diff
@@ -46,8 +46,15 @@ pub struct VaultKey {
 
 /// Keys directory
 pub trait KeyDirectory: Send + Sync {
+	/// Read keys from directory with password
+	#[allow(unused_variables)]
+	fn load_with_password(&self, password: &Option<Password>) -> Result<Vec<SafeAccount>, Error> {
+		unimplemented!()
```
Comment (debris) [inline ethstore/src/accounts_dir/mod.rs]: same issue. badly designed interface

File: ethstore/src/json/crypto.rs [hunk 4]
```diff
@@ -81,7 +82,7 @@ impl<'a> Visitor<'a> for CryptoFieldVisitor {
 			"kdf" => Ok(CryptoField::Kdf),
 			"kdfparams" => Ok(CryptoField::KdfParams),
 			"mac" => Ok(CryptoField::Mac),
-			_ => Err(Error::custom(format!("Unknown field: '{}'", value))),
+			_ => Ok(CryptoField::Ignored),
```
Comment (debris) [inline ethstore/src/json/crypto.rs]: I don't agree with this change. There is definitely a finite number of special cases. Please handle all of them and describe why they exist

File: ethstore/src/json/key_file.rs [hunk 5]
```diff
@@ -46,7 +46,7 @@ pub struct KeyFile {
 	pub id: Uuid,
 	pub version: Version,
 	pub crypto: Crypto,
-	pub address: H160,
+	pub address: Option<H160>,
```
Comment (debris) [inline ethstore/src/json/key_file.rs]: ~I'm not sure about this change either. I don't know how this affects runtime. Without `address` field we don't know that what account we just loaded and I am afraid that it may be simply unusable.~

nvm, I see that you load the password in `from_file`. Please describe in the code why this field is optional.

File: ethstore/src/account/safe_account.rs [hunk 6]
```diff
@@ -76,17 +76,28 @@ impl SafeAccount {
...
 	/// file, the `filename` should be `Some` name. If it is as yet anonymous, then it
-	/// can be left `None`.
-	pub fn from_file(json: json::KeyFile, filename: Option<String>) -> Self {
-		SafeAccount {
+	/// can be left `None`. `password` is used for reading keystore files w/o address.
+	pub fn from_file(json: json::KeyFile, filename: Option<String>, password: &Option<Password>) -> Result<Self, Error> {
+		let crypto = Crypto::from(json.crypto);
+		let address = match json.address {
+			Some(address) => address.into(),
+			None => match password {
```
Comment (debris) [inline ethstore/src/account/safe_account.rs]: what if the json file contains `address` and we provided a password? imo, this should result in an `Error`

File: ethstore/src/account/kdf.rs [hunk 7]
```diff
@@ -26,7 +26,7 @@ pub struct Pbkdf2 {
 	pub c: u32,
 	pub dklen: u32,
 	pub prf: Prf,
-	pub salt: [u8; 32],
+	pub salt: Vec<u8>,
```
Comment (debris) [inline ethstore/src/account/kdf.rs]: why did you change it to a `Vec<u8>`? I see no reason to do that

File: ethstore/src/json/bytes.rs [hunk 8]
```diff
@@ -19,7 +19,7 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
 use serde::de::Error;
 use rustc_hex::{ToHex, FromHex, FromHexError};
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
```
Comment (debris) [inline ethstore/src/json/bytes.rs]: why did you add `derive(Clone)`?

File: ethstore/src/json/kdf.rs [hunk 9]
```diff
@@ -111,7 +111,7 @@ pub struct Pbkdf2 {
 	pub c: u32,
 	pub dklen: u32,
 	pub prf: Prf,
-	pub salt: H256,
+	pub salt: Bytes,
```
Comment (debris) [inline ethstore/src/json/kdf.rs]: why is it changed to `Bytes`?

File: ethkey/cli/src/main.rs [hunk 10]
```diff
@@ -168,7 +168,7 @@ fn main() {
 		Ok(ok) => println!("{}", ok),
 		Err(Error::Docopt(ref e)) => e.exit(),
 		Err(err) => {
-			println!("{}", err);
+			error!("{}", err);
```
[the remark this answers is not in the document]
Comment (tworec) [reply ethkey/cli/src/main.rs]: :+1:

File: ethstore/cli/src/main.rs [hunk 11]
```diff
@@ -146,12 +149,13 @@ impl fmt::Display for Error {
...
 fn main() {
 	panic_hook::set_abort();
+	env_logger::try_init().expect("Logger initialized only once.");
 
 	match execute(env::args()) {
 		Ok(result) => println!("{}", result),
 		Err(Error::Docopt(ref e)) => e.exit(),
 		Err(err) => {
-			println!("{}", err);
+			error!("{}", err);
```
[the remark this answers is not in the document]
Comment (tworec) [reply ethstore/cli/src/main.rs]: :+1:

File: ethstore/cli/src/main.rs [hunk 12]
```diff
@@ -241,7 +245,8 @@ fn execute<S, I>(command: I) -> Result<String, Error> where I: IntoIterator<Item
 	} else if args.cmd_import {
 		let src = key_dir(&args.flag_src)?;
 		let dst = key_dir(&args.flag_dir)?;
-		let accounts = import_accounts(&*src, &*dst)?;
+		let password = load_password(&args.arg_password).ok();
```
[the remark this answers is not in the document]
Comment (tworec) [reply ethstore/cli/src/main.rs]: yup, empty string.

error will be printed when there will be a keystore to import w/o address. Otherwise I would silently ignore invalid pwd path.

File: ethstore/src/account/safe_account.rs (hunk 6 above)
Comment (tworec) [reply ethstore/src/account/safe_account.rs]: Again, I would silently ignore this. As for now all errors reading kestores are silently ignored, which is confusing me too, but to be consistent I propose to ignore it.

File: ethstore/src/json/bytes.rs (hunk 8 above)
Comment (tworec) [reply ethstore/src/json/bytes.rs]: its probably aftermath my debugging... sorry

File: ethstore/src/json/kdf.rs (hunk 9 above)
Comment (tworec) [reply ethstore/src/json/kdf.rs]: again, it's bc we do not know length of salt

File: ethstore/src/json/crypto.rs (hunk 4 above)
Comment (tworec) [reply ethstore/src/json/crypto.rs]: :+1:

File: ethstore/src/accounts_dir/disk.rs (hunk 2 above)
Comment (tworec) [reply ethstore/src/accounts_dir/disk.rs]: thx, will rethink it

**Comment by 5chdn** [comment]
Are you still working on this @tworec

**Comment by tworec** [comment]
I want to finish this. I was looking to see @debris answers...

File: ethstore/src/account/kdf.rs (hunk 7 above)
Comment (niklasad1) [reply ethstore/src/account/kdf.rs]: @tworec just revert the changes

Comment (tworec) [reply ethstore/src/account/kdf.rs]: but we do not the salt length, it can exceed 32 bytes. What then?

Comment (niklasad1) [reply ethstore/src/account/kdf.rs]: All right, sorry did not see your comment there then it makes sense to use a vector instead of an array.

Comment (tworec) [reply ethstore/src/account/kdf.rs]: 👍

File: ethstore/src/account/safe_account.rs (hunk 1 above)
Comment (tworec) [reply ethstore/src/account/safe_account.rs]: fixed

File: ethstore/src/json/crypto.rs (hunk 4 above)
Comment (tworec) [reply ethstore/src/json/crypto.rs]: done

File: ethstore/src/json/crypto.rs [hunk 13]
```diff
@@ -122,6 +124,8 @@ impl<'a> Visitor<'a> for CryptoVisitor {
 				Some(CryptoField::Kdf) => { kdf = Some(visitor.next_value()?); }
 				Some(CryptoField::KdfParams) => { kdfparams = Some(visitor.next_value()?); }
 				Some(CryptoField::Mac) => { mac = Some(visitor.next_value()?); }
+				// skip not required version field (it appears in pyethereum generated keystores)
```
Comment (sorpaas) [inline ethstore/src/json/crypto.rs]: :(

File: ethstore/src/lib.rs [hunk 14]
```diff
@@ -36,7 +36,6 @@ extern crate ethereum_types;
 extern crate ethkey as _ethkey;
 extern crate parity_wordlist;
 
-#[macro_use]
 extern crate log;
```
Comment (sorpaas) [inline ethstore/src/lib.rs]: Are we using `log` crate in `ethstore` at all? If not, maybe remove this altogether?

File: ethstore/cli/src/main.rs [hunk 15]
```diff
@@ -239,8 +239,13 @@ fn execute<S, I>(command: I) -> Result<String, Error> where I: IntoIterator<Item
 			.collect();
 		Ok(format_accounts(&accounts))
 	} else if args.cmd_import {
-		let src = key_dir(&args.flag_src)?;
-		let dst = key_dir(&args.flag_dir)?;
+		let password = match &args.arg_password.as_ref() {
+			&"" => None,
```
Comment (niklasad1) [inline ethstore/cli/src/main.rs]: You can remove `&` here 
```suggestion
			"" => None,
```

File: ethstore/cli/src/main.rs [hunk 16]
```diff
@@ -239,8 +239,13 @@ fn execute<S, I>(command: I) -> Result<String, Error> where I: IntoIterator<Item
 			.collect();
 		Ok(format_accounts(&accounts))
 	} else if args.cmd_import {
-		let src = key_dir(&args.flag_src)?;
-		let dst = key_dir(&args.flag_dir)?;
+		let password = match &args.arg_password.as_ref() {
```
Comment (niklasad1) [inline ethstore/cli/src/main.rs]:
```suggestion
		let password = match args.arg_password.as_ref() {
```

File: ethstore/cli/src/main.rs [hunk 17]
```diff
@@ -239,8 +239,13 @@ fn execute<S, I>(command: I) -> Result<String, Error> where I: IntoIterator<Item
...
+			&"" => None,
+			_ => Some(load_password(&args.arg_password)?)
```
Comment (niklasad1) [inline ethstore/cli/src/main.rs]: I don't like that you wrap potential errors in Option, use `ok()` instead to convert the result to an `Option`

Comment (niklasad1) [inline ethstore/cli/src/main.rs]:
```suggestion
			_ => load_password(&args.arg_password).ok()
```

File: ethstore/src/accounts_dir/disk.rs [hunk 18]
```diff
@@ -117,16 +119,30 @@ pub struct DiskDirectory<T> where T: KeyFileManager {
...
+	}
+}
 
 impl RootDiskDirectory {
 	pub fn create<P>(path: P) -> Result<Self, Error> where P: AsRef<Path> {
 		fs::create_dir_all(&path)?;
 		Ok(Self::at(path))
 	}
 
+	pub fn with_password(&self, password: Option<Password>) -> Self {
```
Comment (niklasad1) [inline ethstore/src/accounts_dir/disk.rs]: Add documentation here, please!

File: ethstore/src/accounts_dir/disk.rs [hunk 19]
```diff
@@ -191,7 +207,7 @@ impl<T> DiskDirectory<T> where T: KeyFileManager {
 					.map_err(Into::into)
 					.and_then(|file| self.key_manager.read(filename, file))
 					.map_err(|err| {
-						warn!("Invalid key file: {:?} ({})", path, err);
+						eprintln!("Invalid key file: {:?} ({})", path, err);
```
Comment (niklasad1) [inline ethstore/src/accounts_dir/disk.rs]: hrm, why `eprintln` here instead of rustlog? I believe we want it to be rustlog

File: ethstore/src/accounts_dir/disk.rs [hunk 20]
```diff
@@ -117,16 +119,30 @@ pub struct DiskDirectory<T> where T: KeyFileManager {
...
 
 /// Keys file manager for root keys directory
-pub struct DiskKeyFileManager;
+pub struct DiskKeyFileManager {
+	password: Option<Password>,
+}
+
+impl DiskKeyFileManager {
+	/// Create new disk key manager w/o password
+	pub fn new() -> Self {
```
Comment (niklasad1) [inline ethstore/src/accounts_dir/disk.rs]: You could derive this impl instead with a default implementation which I prefer in this context!

File: ethstore/src/lib.rs (hunk 14 above)
Comment (niklasad1) [reply ethstore/src/lib.rs]: No, not after the change in this PR:  https://github.com/paritytech/parity-ethereum/blob/43c12d676d4b5fad7b4964d5d55b477087f33c86/ethstore/src/accounts_dir/disk.rs#L210

Comment (sorpaas) [reply ethstore/src/lib.rs]: @niklasad1 `eprintln!` is from `std`. I see that `macro_use` is completely removed so we aren't using any logging macros from `log` crate?

Comment (sorpaas) [reply ethstore/src/lib.rs]: Ah I see, you meant that we should change back to use `log` macros instead for those `eprintln`.

File: ethstore/src/json/crypto.rs (hunk 13 above)
Comment (tworec) [reply ethstore/src/json/crypto.rs]: any constructive criticism @sorpaas? :)

File: ethstore/src/accounts_dir/disk.rs (hunk 19 above)
Comment (tworec) [reply ethstore/src/accounts_dir/disk.rs]: because otherwise, the error message is swallowed by `.ok()` 4 lines below, and if you are running `ethstore/cli` user has no idea what went wrong since there is no logging enabled...

File: ethstore/src/lib.rs (hunk 14 above)
Comment (tworec) [reply ethstore/src/lib.rs]: I also opt to remove `log` crate since it is not used any more. Unless you want me to restore `warn!` where I've added `eprintln`

**Comment by 5chdn** [comment]
:+1: needs a 2nd review

File: ethstore/src/accounts_dir/disk.rs (hunk 19 above)
Comment (niklasad1) [reply ethstore/src/accounts_dir/disk.rs]: Ok, I see but not the `ok()'s` fault IMO. The method is badly designed for your purpose because it only propagates the error if the  `Vec<PathBuf>` fails.

/cc  @sorpaas @debris are we concerned by the performance penalty that `eprintln!` comes with in this context? I'm not exactly sure when this is code is used in the `core client` but I guess it is when we sign transactions and similar?

Comment (tworec) [reply ethstore/src/accounts_dir/disk.rs]: I haven't expected that keyfiles may be read upon every sign (ECC) ops... some dive into parity code has not confirmed this, but if it is the case then I should restore `warn!`.

My first version of this enables logging in `ethstore/cli`. This is also acceptable to me.

**Comment by tworec** [comment]
@niklasad1 done, but I've initialized log level to `warn` also. It will enable logs for invalid password.

This is sample CLI output with invalid password
```
parity-ethereum/ethstore/cli pyethereum-keystore-support (0282536a4)$ cargo run import pwd.txx --src ~/ethstore/
   Compiling ethstore-cli v0.1.1 (/Users/tworec/git/parity-ethereum/ethstore/cli)                                                                                                    
    Finished dev [unoptimized + debuginfo] target(s) in 3.30s                                                                                                                        
     Running `/Users/tworec/git/parity-ethereum/target/debug/ethstore import pwd.txx --src /Users/tworec/ethstore/`
 WARN 2018-12-19T11:05:15Z: ethstore::accounts_dir::disk: Invalid key file: "/Users/tworec/ethstore/keystore-mainnet-backup.json" (Invalid password)
 WARN 2018-12-19T11:05:15Z: ethstore::accounts_dir::disk: Invalid key file: "/Users/tworec/ethstore/mainnet_keystore.json" (Invalid password)
```

**Comment by 5chdn** [comment]
@tworec could you rebase this?

@niklasad1 could you  review this again?

**Comment by tworec** [comment]
@5chdn I think rebase is overkill here. I've merged instead.

**Comment by 5chdn** [comment]
yes, whatever you prefer :)

**Comment by tworec** [comment]
So why do we need a review? Merge introduced no changes in this PR

**Comment by 5chdn** [comment]
yes. one more review. please wait till this holiday craze is over.

**Comment by tworec** [comment]
ok

File: accounts/ethstore/src/accounts_dir/disk.rs [hunk 21]
```diff
@@ -117,16 +119,25 @@ pub struct DiskDirectory<T> where T: KeyFileManager {
...
+}
 
 impl RootDiskDirectory {
 	pub fn create<P>(path: P) -> Result<Self, Error> where P: AsRef<Path> {
 		fs::create_dir_all(&path)?;
 		Ok(Self::at(path))
 	}
 
+	/// allows to read keyfiles with given password (needed for keyfiles w/o address)
+	pub fn with_password(&self, password: Option<Password>) -> Self {
```
Comment (tomusdrw) [inline accounts/ethstore/src/accounts_dir/disk.rs]: I don't think it's a good idea to implement that here. Reasons:
1. The assumption is that all accounts inside the directory share the same password which is most likely incorrect.
2. It's used only for python keyfiles import, but will polute (already convoluted) codebase

Can we not use `DiskDirectory` for python keyfiles import and just allow importing particular account files?
Alternatively can't we treat python keyfiles as a vault (they share the same password actually) and use structs from this code part?

File: accounts/ethstore/src/account/safe_account.rs [hunk 22]
```diff
@@ -77,16 +77,32 @@ impl SafeAccount {
...
-		SafeAccount {
+	pub fn from_file(json: json::KeyFile, filename: Option<String>) -> Result<Self, Error> {
+		SafeAccount::from_file_with_password(json, filename, &None)
+	}
+
+	/// Same as [`from_file`](#method.from_file), but with `password` which enables reading keystore files w/o address.
+	pub fn from_file_with_password(json: json::KeyFile, filename: Option<String>, password: &Option<Password>) -> Result<Self, Error> {
+		let crypto = Crypto::from(json.crypto);
+		let address = match json.address {
+			Some(address) => address.into(),
```
Comment (tomusdrw) [inline accounts/ethstore/src/account/safe_account.rs]: imho if the `password` is provided we should validate that the `adress` is correct. Maybe switch the order here: first check if we have password and after deriving the address optionally compare it with the one in `json`.

File: accounts/ethstore/src/account/safe_account.rs [hunk 23]
```diff
@@ -77,16 +77,32 @@ impl SafeAccount {
 	/// Create a new `SafeAccount` from the given `json`; if it was read from a
 	/// file, the `filename` should be `Some` name. If it is as yet anonymous, then it
 	/// can be left `None`.
-	pub fn from_file(json: json::KeyFile, filename: Option<String>) -> Self {
-		SafeAccount {
+	pub fn from_file(json: json::KeyFile, filename: Option<String>) -> Result<Self, Error> {
```
Comment (tomusdrw) [inline accounts/ethstore/src/account/safe_account.rs]: This now errors for no good reason, afaict it will never fail. Can we please get the proper return type back?

File: accounts/ethstore/src/account/safe_account.rs [hunk 24]
```diff
@@ -77,16 +77,32 @@ impl SafeAccount {
...
+		SafeAccount::from_file_with_password(json, filename, &None)
+	}
+
+	/// Same as [`from_file`](#method.from_file), but with `password` which enables reading keystore files w/o address.
+	pub fn from_file_with_password(json: json::KeyFile, filename: Option<String>, password: &Option<Password>) -> Result<Self, Error> {
```
Comment (tomusdrw) [inline accounts/ethstore/src/account/safe_account.rs]: I would make `password` mandatory here (see above comment). The decision which function to use and what errors to handle can be left to the caller.

File: accounts/ethstore/cli/src/main.rs [hunk 25]
```diff
@@ -146,30 +148,34 @@ impl fmt::Display for Error {
 
 fn main() {
 	panic_hook::set_abort();
+	if env::var("RUST_LOG").is_err() {
+		env::set_var("RUST_LOG", "warn")
```
[the remark this answers is not in the document]
Comment (tworec) [reply accounts/ethstore/cli/src/main.rs]: nope, see https://docs.rs/env_logger/0.5.0/env_logger/#enabling-logging
> by default all logging is disabled except for `error!`

And I want to see [this warning](https://github.com/paritytech/parity-ethereum/pull/9710/files#diff-4b50b928e10747ccbb5b89b25be9e318R205):
`warn!("Invalid key file: {:?} ({})", path, err);`

see discussion above

File: accounts/ethstore/cli/src/main.rs [hunk 26]
```diff
@@ -146,30 +148,34 @@ impl fmt::Display for Error {
...
+	}
+	env_logger::try_init().expect("Logger initialized only once.");
 
 	match execute(env::args()) {
 		Ok(result) => println!("{}", result),
 		Err(Error::Docopt(ref e)) => e.exit(),
 		Err(err) => {
-			println!("{}", err);
+			eprintln!("{}", err);
```
[the remark this answers is not in the document]
Comment (tworec) [reply accounts/ethstore/cli/src/main.rs]: maybe, but see this https://github.com/paritytech/parity-ethereum/pull/9710#discussion_r224011177 by @debris 

guy's pls, can we just proceed and not nitpicking about error messages?

File: accounts/ethstore/src/accounts_dir/disk.rs (hunk 21 above)
Comment (tworec) [reply accounts/ethstore/src/accounts_dir/disk.rs]: ad1. I agree, it is not very generic, but as a user I just create special dir for my keyfiles (even if its single one) before import.

ad 2. it is used for all keyfiles w/o address. Not only pyethereum ones. As any new feature it introduces some noice. 

> Can we not use DiskDirectory for python keyfiles import and just allow importing particular account files?

I've tried to be consistent with current approach which is importing whole directories. Introduction of new `import-single` in CLI seems overkill to me.

>Alternatively can't we treat python keyfiles as a vault (they share the same password actually) and use structs from this code part?

I've analyzed vault's, but it does not fit here. You need to repack your keyfiles into valuts, and in orted to repack you need to import it first, right?

File: accounts/ethstore/src/account/safe_account.rs (hunk 22 above)
Comment (tworec) [reply accounts/ethstore/src/account/safe_account.rs]: nice feature. It will need to throw an error when addresses do not match, but in next comment you're proposing to revert `Result`.

File: accounts/ethstore/src/account/safe_account.rs (hunk 23 above)
Comment (tworec) [reply accounts/ethstore/src/account/safe_account.rs]: it will fail for example when the password is incorrect.

File: accounts/ethstore/src/account/safe_account.rs (hunk 24 above)
Comment (tworec) [reply accounts/ethstore/src/account/safe_account.rs]: sure, it can be done, but for price of code repetition, because I use it to implement `from_file`

File: accounts/ethstore/src/accounts_dir/disk.rs (hunk 21 above)
Comment (tworec) [reply accounts/ethstore/src/accounts_dir/disk.rs]: ad1. if your keyfiles have different pwds you will see warnings, but import will succeed with at least this file with `pwd` you've entered. You can reexecute `import` with another `pwd` to import others.

Comment (tomusdrw) [reply accounts/ethstore/src/accounts_dir/disk.rs]: We have `import-wallet`, but it seems it only works for presale wallets currently. I'd be more inclined to handle it there than to hack around `DiskKeyFileManager` (since the former command allows importing directly to a vault, etc).

That said I'm fine with this if you address the `SafeAccount` semantics in a way I proposed in the other comment.

**Comment by tworec** [comment]
@tomusdrw you're suggestion is incorporated.
@niklasad1 can you pls approve?
